docs/specs/sdk.md

```diff
@@ -0,0 +1,140 @@
+`authgear.foobar()` means calling the `foobar()` method of the container.
+
+Every container has a name.
+If no name is given, the name is `"default"`.
+
+Everything stored by a container is scoped by its name.
+Therefore, if the developer wants to implement multiple account,
+they must create different container instances with distinct names.
+
+## Biometric authentication
+
+Biometric authentication is provided by the following API
+
+```typescript
+interface BiometricKey {
+  userID: string;
+  kid: string;
+  createdAt: Date;
+}
+
+/**
+ * This function can be called at anytime.
+ * This function access the keychain to find out biometric key IDs on this device.
+ * The list of the keys is stored with kSecAttrAccessibleWhenUnlockedThisDeviceOnly,
+ * therefore the data is not included in backup, nor transfered to new device.
+ */
+function listBiometricKeys(): Promise<BiometricKey[]>;
+
+/**
+ * This function can be called when the user has logged in.
+ * This function WILL trigger biometric authentication if necessary.
+ *
+ * 1. An UUID is generated as kid.
+ * 2. A key pair is generated and stored in the keychain.
+ * 3. A challenge is requested from the server.
+ * 4. A Biometric identity is added to the user.
+ *
+ * The key pair is stored with kSecAttrAccessibleWhenPasscodeSetThisDeviceOnly and kSecAccessControlBiometryAny,
+ * therefore the data is not included in backup, nor transfered to new device,
+ * biometric authentication or password code is prompted if necessary.
+ */
+function addBiometric(): Promise<BiometricKey>;
+
+/**
+ * This function can be called when the user has logged in.
+ * This function WILL trigger biometric authentication if necessary.
+ *
+ * 1. The key pair is deleted from the keychain.
+ * 2. If the key pair is not found, save the error as e.
+ * 3. Call /api/identity/biometric/remove
+ * 4. If the identity is not found, throw the error immediately.
+ * 5. If identity was removed and e is non-null, set e to null.
+ * 6. Throw e if it is non-null.
+ *
+ * The key pair is stored with kSecAttrAccessibleWhenPasscodeSetThisDeviceOnly and kSecAccessControlBiometryAny,
+ * therefore the data is not included in backup, nor transfered to new device,
+ * biometric authentication or password code is prompted if necessary.
+ */
+function removeBiometric(kid: string): Promise<void>;
+
+/**
+ * This function can be called at anytime.
+ * This function WILL trigger biometric authentication if necessary.
+ *
+ * 1. The key pair is retrieved from the keychain.
+ * 2. A challenge is requested from the server.
+ * 3. Call /oauth2/token to request refresh token and access token.
+ * 4. If the identity is not found, do the following steps.
+ *   4.1. Remove the kid from the list of keys so that listBiometricKeys no longer includes that kid.
+ */
+function authenticateBiometric(kid: string): Promise<AuthorizeResult>;
+```
+
+### Open questions
+
+#### Should we allow only 1 private key per container name?
```

Comment:
I think it is okay to allow only 1 private key per container, so we won't risk the problem of User A key is accessible to the app even when User B is using

Reply:
@louischan-oursky if we support only 1 private key per container, Seems we don't need the API of `listBiometricKeys()` and `BiometricKey:kid`?

Reply:
Adjusted the API to adapt 1 keypair per container.